src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -85,6 +85,121 @@ public override string ToString()
+
+    C<U> L<U>(U u) => null!;
+}", options: WithNonNullTypesTrue());
+
+            comp.VerifyDiagnostics(
+                // (9,15): warning CS8604: Possible null reference argument for parameter 'index' in 'int C<object>.this[object index]'.
+                //         L(o2)[o] = 3; // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "o").WithArguments("index", "int C<object>.this[object index]").WithLocation(9, 15)
+                );
+        }
+
+        [Fact, WorkItem(29964, "https://github.com/dotnet/roslyn/issues/29964")]
+        public void IndexerUpdatedBasedOnReceiver_GetterArguments()
+        {
+            var comp = CreateCompilation(@"
+class C<T>
+{
+    void M(object? o, object o2)
+    {
+        _ = L(o)[o];
+        _ = L(o)[o2];
+
+        _ = L(o2)[o]; // warn
+        _ = L(o2)[o2];
+    }
+
+    int this[T index] { get => throw null!; set => throw null!; }
+
+    C<U> L<U>(U u) => null!;
+}", options: WithNonNullTypesTrue());
+
+            comp.VerifyDiagnostics(
+                // (9,19): warning CS8604: Possible null reference argument for parameter 'index' in 'int C<object>.this[object index]'.
+                //         _ = L(o2)[o]; // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "o").WithArguments("index", "int C<object>.this[object index]").WithLocation(9, 19)
+                );
+        }
+
+        [Fact, WorkItem(29964, "https://github.com/dotnet/roslyn/issues/29964")]
+        public void IndexerUpdatedBasedOnReceiver_SetterArguments_NestedNullability()
+        {
+            var comp = CreateCompilation(@"
+class C<T>
+{
+    void M(object? o, object o2)
+    {
+        L(o)[L(o)] = 1;
+        L(o)[L(o2)] = 2; // 1
+
+        L(o2)[L(o)] = 3; // 2
+        L(o2)[L(o2)] = 4;
+    }
+
+    int this[C<T> index] { get => throw null!; set => throw null!; }
+
+    C<U> L<U>(U u) => null!;
+}", options: WithNonNullTypesTrue());
+
+            comp.VerifyDiagnostics(
+                // (7,14): warning CS8620: Argument of type 'C<object>' cannot be used for parameter 'index' of type 'C<object?>' in 'int C<object?>.this[C<object?> index]' due to differences in the nullability of reference types.
+                //         L(o)[L(o2)] = 2; // 1
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "L(o2)").WithArguments("C<object>", "C<object?>", "index", "int C<object?>.this[C<object?> index]").WithLocation(7, 14),
+                // (9,15): warning CS8620: Argument of type 'C<object?>' cannot be used for parameter 'index' of type 'C<object>' in 'int C<object>.this[C<object> index]' due to differences in the nullability of reference types.
+                //         L(o2)[L(o)] = 3; // 2
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "L(o)").WithArguments("C<object?>", "C<object>", "index", "int C<object>.this[C<object> index]").WithLocation(9, 15)
+                );
+        }
```

Comment:
>} [](start = 8, length = 1)

Are we testing cases where the top-level nullability is different from the declared nullability?

For instance:
```C#
static void M(object? x, object? y)
{
    if (x == null) return;
    _ = L(x)[y]; // warn
}
```

Reply:
Please test nested nullability as well:
```C#
static void M(object? x)
{
    var y = L(new[] { x });
    y[0][0].ToString(); // warning
    if (x == null) return;
    var z = L(new[] { x });
    z[0][0].ToString(); // ok
}
```

---
In reply to: [266077488](https://github.com/dotnet/roslyn/pull/34134#discussion_r266077488) [](ancestors = 266077488)